### PR TITLE
[#7217] check card state before installation

### LINKED ISSUE #7217: Check card state before installation
# Problem

Card state should be checked before starting the installation. For example, it might already have account on it or there is no slots to pair new device so proper screen with possible actions and explanation should be shown to the user.

# Implementation

Card states diagram:
https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Hardwallet-light-flows?node-id=0%3A1

<blockquote><div> Figma</div><div><strong><a href="https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Hardwallet-light-flows">Figma</a></strong></div><div>Created with Figma</div></blockquote>

### PR DESCRIPTION
**Pull request by dmitryn**
fixes #7217 

### Summary:

Before card setup, installer checks card state which can be:
* `blank` (no applet installed on keycard)
* `pre-init` (applet installed, not initialized)
* `init` (card initialized, but no account on it)
* `account` (card has account/keys on it)
* `not-paired` (card has pairing slots, but not paired to current device)
* `no-pairing-slots` (card as no pairing slots available (5))

Also, we now allow user to set custom PIN during setup.

### Testing notes (optional):

Installation process needs to be tested with the card in all possible states (blank, pre-init, init, account, not-paired, no-pairing-slots). Ping me if you need assistance with putting card in one of the states.

#### Platforms (optional)
- Android

status: ready

### CODE REVIEW

File: src/status_im/ui/screens/hardwallet/components.cljs [hunk 1]
```diff
@@ -97,9 +100,7 @@
       :display-name        "maintain-card"
       :reagent-render      (fn [step] [react/view styles/maintain-card-container
                                        [react/touchable-highlight
```
Comment (jeluard) [inline src/status_im/ui/screens/hardwallet/components.cljs]: Remove?

**Review by jeluard — APPROVED** [review]

File: src/status_im/ui/screens/hardwallet/components.cljs (hunk 1 above)
Comment (dmitryn) [reply src/status_im/ui/screens/hardwallet/components.cljs]: yes, i'll remove it

**Review by flexsurfer — APPROVED** [review]

**Review by jakubgs — APPROVED** [review]
Approving `mobile_files/package.json.orig` changes.

**Comment by asemiankevich** [comment]
@gravityblast Andrea, i am receiving `ERROR[02-14|16:22:31.010] installation error                       package=status-go/cmd/hardware-wallet-light error="unexpected response from command load 0 of 36: 6424"`

could you please assist? it is blocking me from testing

```
DEBUG[02-14|16:22:30.392] using reader                             package=status-go/cmd/hardware-wallet-light name="ACS ACR 38U-CCID"
DEBUG[02-14|16:22:30.393] connecting to card                       package=status-go/cmd/hardware-wallet-light reader="ACS ACR 38U-CCID"
DEBUG[02-14|16:22:30.395] card protocol                            package=status-go/cmd/hardware-wallet-light T=1
installation can take a while...
DEBUG[02-14|16:22:30.395] apdu command                             package=hardware-wallet-go/globalplatform   hex="00 A4 04 00 0F 53 74 61 74 75 73 57 61 6C 6C 65 74 41 70 70"
DEBUG[02-14|16:22:30.411] apdu response                            package=hardware-wallet-go/globalplatform   hex="6A 82"
DEBUG[02-14|16:22:30.411] sending apdu command                     package=status-go/cmd/hardware-wallet-light name=select
DEBUG[02-14|16:22:30.411] apdu command                             package=hardware-wallet-go/globalplatform   hex="00 A4 04 00 08 A0 00 00 01 51 00 00 00"
DEBUG[02-14|16:22:30.424] apdu response                            package=hardware-wallet-go/globalplatform   hex="90 00"
DEBUG[02-14|16:22:30.425] sending apdu command                     package=status-go/cmd/hardware-wallet-light name="initialize update"
DEBUG[02-14|16:22:30.425] apdu command                             package=hardware-wallet-go/globalplatform   hex="80 50 00 00 08 C5 14 04 9A 5E 2D 0A 52 00"
DEBUG[02-14|16:22:30.484] apdu response                            package=hardware-wallet-go/globalplatform   hex="00 00 72 01 00 46 79 07 24 13 FF 02 00 2C F2 EB 6A AB 2C 79 D7 13 5C 6F D8 33 3B CC 90 00"
DEBUG[02-14|16:22:30.485] sending apdu command                     package=status-go/cmd/hardware-wallet-light name="external authenticate"
DEBUG[02-14|16:22:30.485] wrapping apdu command                    package=hardware-wallet-go/globalplatform   hex="84 82 01 00 08 95 6C EB 45 B2 94 A9 0A"
DEBUG[02-14|16:22:30.485] apdu command                             package=hardware-wallet-go/globalplatform   hex="84 82 01 00 10 95 6C EB 45 B2 94 A9 0A 44 A6 5C 1B 00 6C 4E 20"
DEBUG[02-14|16:22:30.522] apdu response                            package=hardware-wallet-go/globalplatform   hex="90 00"
DEBUG[02-14|16:22:30.522] sending apdu command                     package=status-go/cmd/hardware-wallet-light name=delete
DEBUG[02-14|16:22:30.522] wrapping apdu command                    package=hardware-wallet-go/globalplatform   hex="80 E4 00 00 09 4F 07 D2 76 00 00 85 01 01"
DEBUG[02-14|16:22:30.522] apdu command                             package=hardware-wallet-go/globalplatform   hex="84 E4 00 00 11 4F 07 D2 76 00 00 85 01 01 3A 3C DE 32 A4 3E D3 87"
DEBUG[02-14|16:22:30.558] apdu response                            package=hardware-wallet-go/globalplatform   hex="6A 88"
DEBUG[02-14|16:22:30.558] sending apdu command                     package=status-go/cmd/hardware-wallet-light name=delete
DEBUG[02-14|16:22:30.558] wrapping apdu command                    package=hardware-wallet-go/globalplatform   hex="80 E4 00 00 11 4F 0F 53 74 61 74 75 73 57 61 6C 6C 65 74 41 70 70"
DEBUG[02-14|16:22:30.558] apdu command                             package=hardware-wallet-go/globalplatform   hex="84 E4 00 00 19 4F 0F 53 74 61 74 75 73 57 61 6C 6C 65 74 41 70 70 AB 26 B8 34 EF 02 54 CF"
DEBUG[02-14|16:22:30.595] apdu response                            package=hardware-wallet-go/globalplatform   hex="6A 88"
DEBUG[02-14|16:22:30.595] sending apdu command                     package=status-go/cmd/hardware-wallet-light name=delete
DEBUG[02-14|16:22:30.595] wrapping apdu command                    package=hardware-wallet-go/globalplatform   hex="80 E4 00 00 0E 4F 0C 53 74 61 74 75 73 57 61 6C 6C 65 74"
DEBUG[02-14|16:22:30.595] apdu command                             package=hardware-wallet-go/globalplatform   hex="84 E4 00 00 16 4F 0C 53 74 61 74 75 73 57 61 6C 6C 65 74 40 00 0C A7 5F 0A 11 78"
DEBUG[02-14|16:22:30.631] apdu response                            package=hardware-wallet-go/globalplatform   hex="6A 88"
DEBUG[02-14|16:22:30.631] sending apdu command                     package=status-go/cmd/hardware-wallet-light name="install for load"
DEBUG[02-14|16:22:30.631] wrapping apdu command                    package=hardware-wallet-go/globalplatform   hex="80 E6 02 00 19 0C 53 74 61 74 75 73 57 61 6C 6C 65 74 08 A0 00 00 01 51 00 00 00 00 00 00"
DEBUG[02-14|16:22:30.631] apdu command                             package=hardware-wallet-go/globalplatform   hex="84 E6 02 00 21 0C 53 74 61 74 75 73 57 61 6C 6C 65 74 08 A0 00 00 01 51 00 00 00 00 00 00 7A 4A EA 9C 40 90 94 26"
DEBUG[02-14|16:22:30.676] apdu response                            package=hardware-wallet-go/globalplatform   hex="00 90 00"
DEBUG[02-14|16:22:30.679] sending apdu command                     package=status-go/cmd/hardware-wallet-light name="load 0 of 36"
DEBUG[02-14|16:22:30.679] wrapping apdu command                    package=hardware-wallet-go/globalplatform   hex="80 E8 00 00 F7 C4 82 25 FA 01 00 23 DE CA FF ED 02 02 04 00 02 07 A0 00 00 08 04 00 01 11 69 6D 2F 73 74 61 74 75 73 2F 6B 65 79 63 61 72 64 02 00 21 00 23 00 21 00 17 00 29 03 4E 00 70 1B 83 02 32 03 E8 00 00 08 A4 46 5B 00 16 00 0B 02 07 04 02 00 04 00 29 04 00 01 07 A0 00 00 00 62 00 01 05 01 07 A0 00 00 00 62 01 02 05 01 07 A0 00 00 00 62 02 01 05 01 07 A0 00 00 00 62 01 01 03 00 17 02 08 A0 00 00 08 04 00 01 01 03 AC 08 A0 00 00 08 04 00 01 02 14 59 06 00 70 00 00 00 80 00 09 00 09 01 01 00 04 00 A2 00 CC 00 DD 01 8E 01 A4 00 83 03 19 00 15 07 01 00 00 04 AF 00 83 03 01 FF 00 07 01 00 00 14 BF 00 80 00 03 00 03 01 00 00 04 15 F2 16 42 16 54 16 6B 00 80 00 0C 00 08 01 0E 00 00 17 00 17 1F 17 86 18 41 18 95 19 BE 19 E3 1A 6B 1A F1 1B 06 1B 23"
DEBUG[02-14|16:22:30.679] apdu command                             package=hardware-wallet-go/globalplatform   hex="84 E8 00 00 FF C4 82 25 FA 01 00 23 DE CA FF ED 02 02 04 00 02 07 A0 00 00 08 04 00 01 11 69 6D 2F 73 74 61 74 75 73 2F 6B 65 79 63 61 72 64 02 00 21 00 23 00 21 00 17 00 29 03 4E 00 70 1B 83 02 32 03 E8 00 00 08 A4 46 5B 00 16 00 0B 02 07 04 02 00 04 00 29 04 00 01 07 A0 00 00 00 62 00 01 05 01 07 A0 00 00 00 62 01 02 05 01 07 A0 00 00 00 62 02 01 05 01 07 A0 00 00 00 62 01 01 03 00 17 02 08 A0 00 00 08 04 00 01 01 03 AC 08 A0 00 00 08 04 00 01 02 14 59 06 00 70 00 00 00 80 00 09 00 09 01 01 00 04 00 A2 00 CC 00 DD 01 8E 01 A4 00 83 03 19 00 15 07 01 00 00 04 AF 00 83 03 01 FF 00 07 01 00 00 14 BF 00 80 00 03 00 03 01 00 00 04 15 F2 16 42 16 54 16 6B 00 80 00 0C 00 08 01 0E 00 00 17 00 17 1F 17 86 18 41 18 95 19 BE 19 E3 1A 6B 1A F1 1B 06 1B 23 78 72 88 63 23 7E FA A9"
DEBUG[02-14|16:22:31.010] apdu response                            package=hardware-wallet-go/globalplatform   hex="64 24"
ERROR[02-14|16:22:31.010] installation error                       package=status-go/cmd/hardware-wallet-light error="unexpected response from command load 0 of 36: 6424"
```

when executing 

```
./status-hardware-wallet install -l debug -a keycard.cap
```

here is info output if it can say something

```
DEBUG[02-14|16:23:09.342] using reader                             package=status-go/cmd/hardware-wallet-light name="ACS ACR 38U-CCID"
DEBUG[02-14|16:23:09.343] connecting to card                       package=status-go/cmd/hardware-wallet-light reader="ACS ACR 38U-CCID"
DEBUG[02-14|16:23:09.346] card protocol                            package=status-go/cmd/hardware-wallet-light T=1
DEBUG[02-14|16:23:09.346] apdu command                             package=hardware-wallet-go/globalplatform   hex="00 A4 04 00 0F 53 74 61 74 75 73 57 61 6C 6C 65 74 41 70 70"
DEBUG[02-14|16:23:09.361] apdu response                            package=hardware-wallet-go/globalplatform   hex="6A 82"
Installed: false
Initialized: false
InstanceUID: 0x
PublicKey: 0x
Version: 0x
AvailableSlots: 0x
KeyUID: 0x
```

**Comment by gravityblast** [comment]
@asemiankevich can you tell me which cap file you are using please? so that I can try

**Comment by asemiankevich** [comment]
> @asemiankevich can you tell me which cap file you are using please? so that I can try

last 2 from releases page. `keycard` and `keycard_v2`

**Comment by asemiankevich** [comment]
@dmitryn would you mind to rebase? I tried to trigger new build but it fails, maybe you need to resolve some conflicts or whatever https://ci.status.im/job/status-react/job/prs/job/android/view/change-requests/job/PR-7460/7/

<blockquote><div><strong><a href="https://ci.status.im/job/status-react/job/prs/job/android/view/change-requests/job/PR-7460/7/">status-react » prs » android » PR-7460 #7 [Jenkins]</a></strong></div></blockquote>

**Comment by asemiankevich** [comment]
1. can't get this screen `No Keycard applet on card`
image

Once i run `./status-hardware-wallet delete -l debug` and then hold the card - this screen is shown
image

2. can't get this screen `Card is out of factory`. Applet is installed but not initialized:

```
Installed: true
Initialized: false
InstanceUID: 0x
PublicKey: 0x049afce895a497828ecae33c924c7ab4bf7a1f494eac0378868a5bdf2537badb71eae5b73b7b6e7eedd695c6dd5516764f98f859fbd2d7c65737ffa10456171824
Version: 0x
AvailableSlots: 0x
KeyUID: 0x
```

I receive basic screen
 image

@dmitryn maybe it is something wrong? either i am doing something wrong or maybe using wrong build (using the last successful) or maybe wrong .cap? (keycard.cap is used). Could you please take a look?

**Comment by dmitryn** [comment]
@asemiankevich 

>1. can't get this screen `No Keycard applet on card`

"no applet on card"  (state 0) should lead to that screen, but also alert with corresponding messages should also be shown. I'll check why there is no alert shown.


>2. can't get this screen `Card is out of factory`. Applet is installed but not initialized:

We don't have separate screen for that. In this scenario, we just start with first installation screen (screen from screenshot).

**Comment by dmitryn** [comment]
@asemiankevich please try new build http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190219-162714-b8212c-pr.apk

<blockquote></blockquote>

**Comment by asemiankevich** [comment]
@dmitryn i've noticed that if i enter the Pairing code manually , so for example i installed and initialized applet with card reader and then i hold the card to the phone -> pairing screen shown -> prompt shown to enter the pairing key, so when i enter the key - i always receive `something went wrong`

image


Could you please check on your side? seems this manual input is not working. Tried a lot of times.

**Comment by dmitryn** [comment]
@asemiankevich sure, i will check it

**Comment by dmitryn** [comment]
@asemiankevich here is new build, please take a look http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190220-155604-51545d-pr.apk

Pairing screen has been removed from the flow because it doesn't make sense in current setup.
If card is blank or out of factory, we will pair card during installation. 
If card has account on it, there is no point in pairing because you still cannot use paired card as no corresponding account exist in Status). You need first remove existing account from card, then proceed with setup with empty card.

<blockquote></blockquote>

**Comment by asemiankevich** [comment]
@dmitryn 
i have already applet on the card installed and intialized. When  i use the card and hold to the phone on [Create new account] flow -> hint to login and reset card first is shown then i tap OK and then setting up PIN screen is shown and at the end it stuck on Preparing card. 

So question: if we force user to login and reset card before new account is created within the card:
- can we block further flow so when i Tap OK -> nothing will happen?
- what if i can't login anymore to reset the card (unistalled the status app, whatever) - this card seems to be unusable?

Attaching some screens below.

image

image

image

**Comment by dmitryn** [comment]
>* what if i can't login anymore to reset the card (unistalled the status app, whatever) - this card seems to be unusable?

@asemiankevich Yes. There is no way to install new keys on the card that already has keys. And if user cannot reset card, it is effectively useless. 

Keycard flows diagram shows this ("State 3" doesn't allow user to proceed) 
screenshot 2019-02-21 19 41 57


Instead, we can ask if user really wants to proceed with installation which will remove current keys from card. @denis-sharypin what do you think?

**Comment by denis-sharypin** [comment]
Idk, so technically I can take your card, reset it and make it mine without providing any codes. It doesn't sound good. @guylouis what do you think?

**Comment by guylouis** [comment]
I agree with @denis-sharypin remark
Only someone who knows the PIN of the card (and thus can login) should be able to wipe out the card of its secrets (and indeed you need Status app installed to do so). So I think the current flow is ok.

**Comment by dmitryn** [comment]
Thanks for feedback guys! 

@asemiankevich i will make changes we discussed and let you know when new build is ready.

* show "enter pair code" screen when card has account but not paired
* remove "unpair" option from keycard settings after card was unpaired
* do not allow user to proceed installation with card in State 3

**Comment by dmitryn** [comment]
@asemiankevich please check new build http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190222-183758-4f7d3a-pr.apk

<blockquote></blockquote>

**Comment by asemiankevich** [comment]
thanks @dmitryn the only thing i think we can do better here -> when i bypass pairing screen -> i receive this. Same alert is shown when i try to create account with card in State 3 (which is okay). I think we can hide this alert when i paired the card. But this could be done later for sure. @guylouis 
image

**Comment by dmitryn** [comment]
@asemiankevich thanks for testing! 
Alert is shown because after card is paired, we end up in State 3 (card paired and has account on it).

**Pull request merged**